environs/manual/winrmprovisioner/winrmprovisioner.go

```diff
@@ -0,0 +1,581 @@
+		Jobs:                    []multiwatcher.MachineJob{multiwatcher.JobHostUnits},
+	}
+	return machineParams, nil
+}
+
+// checkProvisioned checks if the machine is already provisioned or not.
+// if it's already provisioned it will return true
+func checkProvisioned(host string, cli manual.WinrmClientAPI) (bool, error) {
+	logger.Infof("Checking if %s windows machine is already provisioned", host)
+	var stdout, stderr bytes.Buffer
+
+	// run the command trough winrm
+	// this script detects if the jujud process, service is up and running
+	script, err := shell.NewPSEncodedCommand(detectJujudProcess)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	// send the script to the windows machine
+	if err = cli.Run(script, &stdout, &stderr); err != nil {
+		return false, errors.Trace(err)
+	}
+
+	provisioned := strings.Contains(stdout.String(), "Yes")
+	if stderr.Len() != 0 {
+		err = fmt.Errorf("%v (%v)", err, strings.TrimSpace(stderr.String()))
+	}
+
+	// if the script said yes
+	if provisioned {
+		logger.Infof("%s is already provisioned", host)
+	} else {
+		logger.Infof("%s is not provisioned", host)
+	}
+
+	return provisioned, err
+}
+
+// DetectSeriesAndHardwareCharacteristics detects the windows OS
+// series and hardware characteristics of the remote machine
+// by connecting to the machine and executing a bash script.
+func DetectSeriesAndHardwareCharacteristics(host string, cli manual.WinrmClientAPI) (hc instance.HardwareCharacteristics, series string, err error) {
+	logger.Infof("Detecting series and characteristics on %s windows machine", host)
+	var stdout, stderr bytes.Buffer
+
+	script, err := newDetectHardwareScript()
+	if err != nil {
+		return hc, "", err
+	}
+
+	// send the script to the windows machine
+	if err = cli.Run(script, &stdout, &stderr); err != nil {
+		return hc, "", errors.Trace(err)
+	}
+
+	if stderr.Len() != 0 {
+		return hc, "", fmt.Errorf("%v (%v)", err, strings.TrimSpace(stderr.String()))
+	}
+
+	info, err := splitHardWareScript(stdout.String())
+	if err != nil {
+		return hc, "", errors.Trace(err)
+	}
+
+	series = strings.Replace(info[2], "\r", "", -1)
+
+	if err = initHC(&hc, info); err != nil {
+		return hc, "", errors.Trace(err)
+	}
+
+	return hc, series, nil
+}
+
+// initHC it will initialize the hardware characterisrics struct with the
+// parsed and checked info slice string
+// info description :
+//  - info[0] the arch of the machine
+//  - info[1] the ammount of memory that the machine has
+//  - info[2] the series of the machine
+//  - info[3] the number of cores that the machine has
+// It returns nil if it parsed successfully.
+func initHC(hc *instance.HardwareCharacteristics, info []string) error {
+	// add arch
+	arch := arch.NormaliseArch(info[0])
+	hc.Arch = &arch
+
+	// parse the mem number
+	mem, err := strconv.ParseUint(info[1], 10, 64)
+	if err != nil {
+		return errors.Annotatef(err, "Can't parse mem number of the windows machine")
+	}
+	hc.Mem = new(uint64)
+	*hc.Mem = mem
```

Comment:
> No, because after we exit the scope the hc.Mem will be nil.

"mem" will be captured by the reference, and will be stored on the heap (assuming hc is also stored on the heap). It's fine (and preferable) to do  `hc.Mem = &mem`.